src/nupic/algorithms/Segment.cpp

```diff
@@ -392,7 +389,7 @@ void Segment::freeNSynapses(UInt numToFree,
   }
 
   // Now sort the list of candidate synapses
-  std::sort(candidates.begin(), candidates.end(), InPermanenceOrder());
+  std::stable_sort(candidates.begin(), candidates.end(), InPermanenceOrder());
```

Comment:
Realistically, with a wee bit of refactoring I think this operation can be done without the sort altogether.